File: web/src/models/Service.js
/**
 * Model of a service.
 */
export class Service {
    /**
     * @constructor
     * @param {object} service
     */
    constructor(service) {
        this.name = service.name || "";
        this.icon = service.icon || "";
        this.url = service.url || "";
        this.groups = service.groups || [];
        this.category = service.category || 1000;
        this.labels = service.labels || [];
    }

    /**
     * Returns the list of services as an array of Service
     * objects from the given raw services data array.
     * @param {Service[]} serviceDatas 
     * @returns Service[]
     */
    static map(serviceDatas) {
        const services = serviceDatas.map((service) => {
            return new Service(service);
        });
        return services;
    }
}

File: internal/routine/service.go
package routine

import (
	"html/template"

	"github.com/r7wx/easy-gate/internal/config"
	"github.com/r7wx/easy-gate/internal/service"
)

func (r *Routine) getServices(cfg *config.Config) []service.Service {
	type serviceWrapper struct {
		service service.Service
		index   int
	}

	servicePChan := make(chan serviceWrapper)
	for index, cfgService := range cfg.Services {
		go func(index int, cfgService config.Service) {
			servicePChan <- serviceWrapper{
				service: service.Service{
					Icon:     r.getIconData(cfgService),
					Name:     cfgService.Name,
					URL:      template.URL(cfgService.URL),
					Category: cfgService.Category,
					Groups:   cfgService.Groups,
				},
				index: index,
			}
		}(index, cfgService)
	}

	processedServices := map[int]service.Service{}
	for i := 1; i <= len(cfg.Services); i++ {
		processedService := <-servicePChan
		processedServices[processedService.index] = processedService.service
	}

	services := []service.Service{}
	for index := range cfg.Services {
		services = append(services,
			processedServices[index])
	}

	return services
}
